app/components/steps/rank.js

```diff
@@ -0,0 +1,372 @@
+
+const minSelectionsTable = {
+  0: { least: 0, most: 0 },
+  1: { least: 0, most: 0 },
+  2: { least: 0, most: 0 },
+  3: { least: 0, most: 1 },
+  4: { least: 1, most: 1 },
+  5: { least: 1, most: 1 },
+  6: { least: 1, most: 2 },
+  7: { least: 1, most: 2 },
+  8: { least: 1, most: 2 },
+  9: { least: 1, most: 2 },
+  10: { least: 1, most: 2 },
+  11: { least: 1, most: 2 },
+  12: { least: 1, most: 2 },
+}
+
+export default function RankStep(props) {
+  const { onDone, ...otherProps } = props
+  const { data, upsert } = useContext(DeliberationContext)
+
+  const args = derivePointRankGroupList(data)
+
+  function handleOnDone({ valid, value, delta }) {
+    if (delta) upsert({ preRankByParentId: { [delta.parentId]: delta } })
+    window.socket.emit('upsert-rank', delta)
+    onDone({ valid, value })
+  }
+
+  // fetch previous data
+  if (typeof window !== 'undefined')
+    useState(() => {
+      const { discussionId, round } = data
+
+      window.socket.emit('get-user-ranks', discussionId, round, 'pre', result => {
+        if (!result) return // there was an error
+        const [ranks] = result
+        const preRankByParentId = ranks.reduce(
+          (preRankByParentId, rank) => ((preRankByParentId[rank.parentId] = rank), preRankByParentId),
+          {}
+        )
+        upsert({ preRankByParentId })
+      })
+    })
+
+  return (
+    <RankPoints {...args} onDone={handleOnDone} discussionId={data.discussionId} round={data.round} {...otherProps} />
+  )
+}
+
+const toRankString = {
+  undefined: '',
+  most: 'Most',
+  least: 'Least',
+  neutral: 'Neutral',
+}
+
+const rankStringToCategory = Object.entries(toRankString).reduce((rS2C, [key, value]) => {
+  if (key === 'undefined') return rS2C
+  rS2C[value] = key
+  return rS2C
+}, {})
+
+export function RankPoints(props) {
+  const classes = useStylesFromThemeFunction()
+  const {
+    className = '', // may or may not be passed. Should be applied to the outer most tag, after local classNames
+    onDone = () => {}, // a function that is called when the button is clicked.  - if it exists
+    pointRankGroupList,
+    round,
+    discussionId,
+    ...otherProps
+  } = props
+
+  if (!pointRankGroupList) return null
+
+  const [rankByParentId, setRankByParentId] = useState(
+    (pointRankGroupList || []).reduce((rankByParentId, rankPoint) => {
+      if (rankPoint.rank && rankByParentId[rankPoint.point._id]) rankByParentId[rankPoint.point._id] = rankPoint.rank
+      return rankByParentId
+    }, {})
+  )
+
+  const [clearRanksBlocked, setClearRanksBlocked] = useState(true)
+
+  useEffect(() => {
+    const newRankByParentId = {}
+    pointRankGroupList?.forEach(rankPoint => {
+      if (rankPoint.rank) {
+        if (clearRanksBlocked) {
+          newRankByParentId[rankPoint.point._id] = rankPoint.rank
+        } else if (newRankByParentId[rankPoint.point._id]) {
+          newRankByParentId[rankPoint.point._id] = { ...rankPoint.rank, category: '' }
```

Comment:
`clearRanksBlocked` fixed the issue where clearing ranks and display the new ranks are mutually exclusive. 

In rerank.js, it's commented that ranks removed from rankByParentId reappear because deleting ranks isn't a use case. I added this case by only re-adding those ranks to newRankByParentId if it's still in the original, and skipping it if else not. 

However, in the initial render (and after every change to the ranks as I found out), rankByParentId isn't fully initialized, so no ranks are rendered from the data and the bubbles are empty. 

That's where clearRanksBlocked came in, which is initialized to true, so every rank in pointRankGroupList is put into rankByParentId and rendered even if it hasn't been ported to rankByParentId yet. I checked again and saw the second condition never ran as you mentioned, so I removed the redundancies and turned it into this. 

```
if (rankPoint.rank && clearRanksBlocked) {
  // add to newRankByParentId, which renders the rank properly
} 
// do nothing if else
```

Then when the Clear All button is clicked, it sets clearRanksBlocked to false, and the above conditional relies solely on `rankPoint.rank` being true (i.e. it exists). If it doesn't, then it isn't added to newRankByParentId, it's not in the data and doesn't get rendered subsequently, and therefore the rank is now cleared. 

clearRanksBlocked is reverted to true immediately after that so the ranks can properly render again from the data.

I'm honestly still a bit confused myself, but that was the solution I arrived at to make them work concurrently.